File: finanze/domain/transactions.py
from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from domain.base import BaseData
from domain.dezimal import Dezimal
from domain.entity import Entity
from domain.fetch_record import DataSource
from domain.global_position import (
    EquityType,
    FundType,
    ProductType,
)
from pydantic.dataclasses import dataclass


class TxType(str, Enum):
    BUY = "BUY"
    SELL = "SELL"
    DIVIDEND = "DIVIDEND"
    RIGHT_ISSUE = "RIGHT_ISSUE"
    RIGHT_SELL = "RIGHT_SELL"
    SUBSCRIPTION = "SUBSCRIPTION"
    SWAP_FROM = "SWAP_FROM"
    SWAP_TO = "SWAP_TO"

    TRANSFER_IN = "TRANSFER_IN"
    TRANSFER_OUT = "TRANSFER_OUT"
    SWITCH_FROM = "SWITCH_FROM"
    SWITCH_TO = "SWITCH_TO"

    INVESTMENT = "INVESTMENT"
    REPAYMENT = "REPAYMENT"
    INTEREST = "INTEREST"

    FEE = "FEE"

    EXPENSE = "EXPENSE"
    INCOME = "INCOME"


OUTGOING_ACCOUNT_TX_TYPES = {
    TxType.EXPENSE,
    TxType.TRANSFER_OUT,
    TxType.FEE,
}


def account_tx_signed_amount(tx_type: "TxType", amount: "Dezimal") -> "Dezimal":
    """Signed balance effect of an account-product transaction: negative for
    outgoing types (expense/transfer-out/fee), positive otherwise (income,
    transfer-in, interest, ...)."""
    return -amount if tx_type in OUTGOING_ACCOUNT_TX_TYPES else amount


@dataclass(kw_only=True)
class BaseTx(BaseData):
    id: Optional[UUID]
    ref: str
    name: str
    amount: Dezimal
    currency: str
    type: TxType
    date: datetime
    entity: Entity
    source: DataSource
    product_type: ProductType
    entity_account_id: Optional[UUID] = None


@dataclass(kw_only=True)
class BaseInvestmentTx(BaseTx):
    pass


@dataclass(kw_only=True)
class AccountTx(BaseTx):
    fees: Dezimal
    retentions: Dezimal
    interest_rate: Optional[Dezimal] = None
    avg_balance: Optional[Dezimal] = None
    net_amount: Optional[Dezimal] = None
    category: Optional[str] = None
    account_name: Optional[str] = None


@dataclass(kw_only=True)
class StockTx(BaseInvestmentTx):
    shares: Dezimal
    price: Dezimal
    fees: Dezimal
    net_amount: Optional[Dezimal] = None
    isin: Optional[str] = None
    ticker: Optional[str] = None
    market: Optional[str] = None
    retentions: Optional[Dezimal] = None
    order_date: Optional[datetime] = None
    linked_tx: Optional[str] = None
    equity_type: Optional[EquityType] = None


@dataclass(kw_only=True)
class CryptoCurrencyTx(BaseInvestmentTx):
    currency_amount: Dezimal
    symbol: str
    price: Dezimal
    fees: Dezimal
    contract_address: Optional[str] = None
    net_amount: Optional[Dezimal] = None
    retentions: Optional[Dezimal] = None
    order_date: Optional[datetime] = None


@dataclass(kw_only=True)
class MarketForecastTx(BaseInvestmentTx):
    symbol: str
    size: Dezimal
    price: Dezimal
    fees: Dezimal
    contract_address: Optional[str] = None
    net_amount: Optional[Dezimal] = None
    retentions: Optional[Dezimal] = None
    order_date: Optional[datetime] = None


@dataclass(kw_only=True)
class FundTx(BaseInvestmentTx):
    shares: Dezimal
    price: Dezimal
    fees: Dezimal
    net_amount: Optional[Dezimal] = None
    isin: Optional[str] = None
    market: Optional[str] = None
    retentions: Optional[Dezimal] = None
    order_date: Optional[datetime] = None
    fund_type: Optional[FundType] = None


@dataclass(kw_only=True)
class FundPortfolioTx(BaseInvestmentTx):
    portfolio_name: str
    iban: Optional[str] = None
    fees: Dezimal = Dezimal(0)


@dataclass(kw_only=True)
class FactoringTx(BaseInvestmentTx):
    fees: Dezimal
    retentions: Dezimal
    net_amount: Optional[Dezimal] = None


@dataclass(kw_only=True)
class RealEstateCFTx(BaseInvestmentTx):
    fees: Dezimal
    retentions: Dezimal
    net_amount: Optional[Dezimal] = None


@dataclass(kw_only=True)
class DepositTx(BaseInvestmentTx):
    fees: Dezimal
    retentions: Dezimal
    net_amount: Optional[Dezimal] = None


@dataclass
class Transactions:
    investment: Optional[list[BaseInvestmentTx]] = None
    account: Optional[list[AccountTx]] = None

    def __add__(self, other):
        investment = (self.investment or []) + (other.investment or [])
        account = (self.account or []) + (other.account or [])
        return Transactions(investment=investment, account=account)


@dataclass
class TransactionsResult:
    transactions: list[BaseTx]


@dataclass
class TransactionQueryRequest:
    page: int = 1
    limit: int = 10
    entities: Optional[list[UUID]] = None
    excluded_entities: Optional[list[UUID]] = None
    product_types: Optional[list[ProductType]] = None
    from_date: Optional[datetime] = None
    to_date: Optional[datetime] = None
    types: Optional[list[TxType]] = None
    historic_entry_id: Optional[UUID] = None


@dataclass
class CategorySpend:
    category: str
    amount: Dezimal
    currency: str


@dataclass
class MonthlySpend:
    month: str
    expense: Dezimal
    income: Dezimal
    currency: str


@dataclass
class CurrencyAmount:
    amount: Dezimal
    currency: str


@dataclass
class SpendingSummaryRequest:
    from_date: Optional[datetime] = None
    to_date: Optional[datetime] = None


@dataclass
class SpendingSummaryResult:
    by_category: list[CategorySpend]
    by_month: list[MonthlySpend]
    total_expense: list[CurrencyAmount]
    total_income: list[CurrencyAmount]


@dataclass
class AccountLedgerEntry:
    transaction: AccountTx
    balance_after: Dezimal


@dataclass
class AccountLedgerResult:
    account_name: str
    currency: str
    current_balance: Dezimal
    entries: list[AccountLedgerEntry]

File: frontend/app/src/types/transactions.ts
import { EquityType, FundType, ProductType } from "./position"
import { DataSource, EntityOrigin } from "."

export enum TxType {
  BUY = "BUY",
  SELL = "SELL",
  DIVIDEND = "DIVIDEND",
  RIGHT_ISSUE = "RIGHT_ISSUE",
  RIGHT_SELL = "RIGHT_SELL",
  SUBSCRIPTION = "SUBSCRIPTION",
  SWAP_FROM = "SWAP_FROM",
  SWAP_TO = "SWAP_TO",

  TRANSFER_IN = "TRANSFER_IN",
  TRANSFER_OUT = "TRANSFER_OUT",
  SWITCH_FROM = "SWITCH_FROM",
  SWITCH_TO = "SWITCH_TO",

  INVESTMENT = "INVESTMENT",
  REPAYMENT = "REPAYMENT",
  INTEREST = "INTEREST",

  FEE = "FEE",

  EXPENSE = "EXPENSE",
  INCOME = "INCOME",
}

// Base transaction interface
export interface BaseTx {
  id: string
  ref: string
  name: string
  amount: number
  currency: string
  type: TxType
  date: string
  entity: {
    id: string
    name: string
    origin: EntityOrigin
  }
  source: DataSource
  product_type: ProductType
}

// Base investment transaction interface
export type BaseInvestmentTx = BaseTx

// Account transaction interface
export interface AccountTx extends BaseTx {
  fees: number
  retentions: number
  interest_rate?: number
  avg_balance?: number
  net_amount?: number
  category?: string
  account_name?: string
}

// Stock transaction interface
export interface StockTx extends BaseInvestmentTx {
  net_amount: number
  isin?: string
  shares: number
  price: number
  fees: number
  ticker?: string
  market?: string
  retentions?: number
  order_date?: string
  linked_tx?: string
  equity_type?: EquityType
}

export interface CryptoCurrencyTx extends BaseInvestmentTx {
  net_amount: number
  currency_amount: number
  price: number
  fees: number
  symbol: string
  contract_address?: string
  retentions?: number
  order_date?: string
}

export interface MarketForecastTx extends BaseInvestmentTx {
  net_amount: number
  size: number
  price: number
  fees: number
  symbol: string
  retentions?: number
  order_date?: string
}

// Fund transaction interface
export interface FundTx extends BaseInvestmentTx {
  net_amount: number
  isin: string
  shares: number
  price: number
  market: string
  fees: number
  retentions?: number
  order_date?: string
  fund_type?: FundType
}

export interface FundPortfolioTx extends BaseInvestmentTx {
  fees: number
  portfolio_name: string
  iban?: string
}

// Factoring transaction interface
export interface FactoringTx extends BaseInvestmentTx {
  net_amount: number
  fees: number
  retentions: number
}

// Real state crowdfunding transaction interface
export interface RealEstateCFTx extends BaseInvestmentTx {
  net_amount: number
  fees: number
  retentions: number
}

// Deposit transaction interface
export interface DepositTx extends BaseInvestmentTx {
  net_amount: number
  fees: number
  retentions: number
}

export interface ManualTransactionBasePayload {
  id: string
  ref: string
  name: string
  amount: number
  currency: string
  type: TxType
  date: string
  entity_id: string
  source: DataSource
  product_type: ProductType
}

export interface ManualAccountTransactionPayload extends ManualTransactionBasePayload {
  product_type: ProductType.ACCOUNT
  fees?: number
  retentions?: number
  interest_rate?: number
  avg_balance?: number
  category?: string
  account_name?: string
}

export interface ManualStockTransactionPayload extends ManualTransactionBasePayload {
  product_type: ProductType.STOCK_ETF
  ticker?: string
  isin?: string
  shares: number
  price: number
  fees?: number
  retentions?: number
  market?: string
  order_date?: string
}

export interface ManualFundTransactionPayload extends ManualTransactionBasePayload {
  product_type: ProductType.FUND
  isin: string
  shares: number
  price: number
  fees?: number
  retentions?: number
  market?: string
  order_date?: string
}

export interface ManualFundPortfolioTransactionPayload extends ManualTransactionBasePayload {
  product_type: ProductType.FUND_PORTFOLIO
  portfolio_name: string
  fees?: number
  iban?: string
}

export interface ManualFactoringTransactionPayload extends ManualTransactionBasePayload {
  product_type: ProductType.FACTORING
  fees?: number
  retentions?: number
}

export interface ManualRealEstateTransactionPayload extends ManualTransactionBasePayload {
  product_type: ProductType.REAL_ESTATE_CF
  fees?: number
  retentions?: number
}

export interface ManualDepositTransactionPayload extends ManualTransactionBasePayload {
  product_type: ProductType.DEPOSIT
  fees?: number
  retentions?: number
}

export interface ManualCryptoCurrencyTransactionPayload extends ManualTransactionBasePayload {
  product_type: ProductType.CRYPTO
  symbol: string
  currency_amount: number
  price: number
  fees?: number
  retentions?: number
  order_date?: string
  contract_address?: string
}

export type ManualTransactionPayload =
  | ManualAccountTransactionPayload
  | ManualStockTransactionPayload
  | ManualFundTransactionPayload
  | ManualFundPortfolioTransactionPayload
  | ManualFactoringTransactionPayload
  | ManualRealEstateTransactionPayload
  | ManualDepositTransactionPayload
  | ManualCryptoCurrencyTransactionPayload

// Transactions container
export interface Transactions {
  investment?: BaseInvestmentTx[]
  account?: AccountTx[]
}

type Tx = AccountTx &
  StockTx &
  CryptoCurrencyTx &
  MarketForecastTx &
  FundTx &
  FactoringTx &
  RealEstateCFTx &
  DepositTx

// Transaction query result
export interface TransactionsResult {
  transactions: Tx[]
}

// Transaction query request
export interface TransactionQueryRequest {
  page?: number
  limit?: number
  entities?: string[]
  product_types?: ProductType[]
  from_date?: string
  to_date?: string
  types?: TxType[]
  historic_entry_id?: string
}

// Spending analysis types
export interface CategorySpend {
  category: string
  amount: number
  currency: string
}

export interface MonthlySpend {
  month: string
  expense: number
  income: number
  currency: string
}

export interface CurrencyAmount {
  amount: number
  currency: string
}

export interface SpendingSummaryResult {
  by_category: CategorySpend[]
  by_month: MonthlySpend[]
  total_expense: CurrencyAmount[]
  total_income: CurrencyAmount[]
}

// Account ledger (running balance) types
export interface AccountLedgerEntry {
  transaction: AccountTx
  balance_after: number
}

export interface AccountLedgerResult {
  account_name: string
  currency: string
  current_balance: number
  entries: AccountLedgerEntry[]
}
